### PR TITLE
Make sorting faster by implementing an insertion sort using an item distance

### LINKED ISSUE #1047: Sorting is slow for duplicate items
**Running Software (issues missing this information will be deleted):**
 - Addon version: 8.2.16 (tried 8.2.19 beta too, but the clean functionality was broken because it tried to put items in my keyring).
 - Server patch: Build 1.13.3.32836

 **Have you read the changelog? (please don't waste our time)**
✅ 

**Describe the bug**
When using the "clean items" functionality, sorting is suboptimal and slow for items you have multiple of. For example all Soul Shards are moved one by one instead of only moving a single Soul Shard.

**To Reproduce**
Steps to reproduce the behaviour:
1. Have multiple similar items in your bag. Can be anything like a Soul Shard or Runecloth.
2. Move the first of the similar items to the bottom of the bag.
3. Click on 'Clean items'
4. Watch each item getting moved one by one down by one slot.

**Expected behaviour**
The last item moved should be moved to the first slot to fill the gap.

**Screenshots**
Here's a video that shows the issue with two kinds of items: https://giphy.com/gifs/fvfe9eXtf2P1PWEA7j

**Error Logs**
No errors.

**Additional context**
It's mainly bothersome for warlocks as soul shards are consumed from the first slot and new ones are placed in the last slow available. However, you can run into the issue with any gathering profession.

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
Fix https://github.com/tullamods/Bagnon/issues/1047. This permit to have a lot fewer item moved by generally choosing the best object to move and not locking the other one. I think the complexity goes from o(n²) to o(nlogn).

The distance is not perfect but is sufficient for most cases in my opinion.

See before : https://media.giphy.com/media/WTKzseVv7BIl7xrIv0/giphy.gif
And after : https://media.giphy.com/media/QYXxVYp9KQ5w4Toi6v/giphy.gif

### CODE REVIEW

**Comment by Pierre-Sassoulas** [comment]
Example of a case where it takes two moves instead of one because the distance is imperfect : https://media.giphy.com/media/J1ub5n76FMM6X1B4Tu/giphy.gif

I think with the position inside the monobag it would be better.

**Comment by Pierre-Sassoulas** [comment]
I just realized it creates a problem with not stopping properly all the time and not stacking properly in some case. Please do not merge 😅

**Comment by Pierre-Sassoulas** [comment]
Ok, I fixed the problems and cleaned the commit tree.

This change makes most cases faster, except when there is only stacking to do which stays the same. Stacking and moving are parallel and trying to make better stacking make moving item less efficient. I think moving is more important as it also stacks object.

Click on links for a video.

# [Before](https://www.youtube.com/watch?v=GPbPCW-Ya38)

## [Moving one stack](https://youtu.be/GPbPCW-Ya38?t=4)

* 2 seconds
* 4 parallels call
* 5 items moved 

## [Moving 2 items out of their stack](https://youtu.be/GPbPCW-Ya38?t=12)
*  4 seconds
* 8 parallels call
* 82 items moved

## [Moving arrows outside of their quiver](https://youtu.be/GPbPCW-Ya38?t=18)
* 11 seconds
* 14 parallels call
* 216 items moved

## [Mixing up 4 stacks](https://youtu.be/GPbPCW-Ya38?t=31)
* 1 second
* 4 parallels call
* 9 items moved

# [After](https://youtu.be/z1543RQFdaw)

## [Moving one stack](https://youtu.be/z1543RQFdaw?t=2)
* Less than a second
* 1 call
* 1 item moved 

## [Moving 2 items out of their stack](https://youtu.be/z1543RQFdaw?t=9)
* 4 seconds
* 8 parallels call
* 71 items moved

## [Moving arrows outside of their quiver](https://youtu.be/z1543RQFdaw?t=16)
* 2 seconds
*  3 parallels call
* 29 items moved

## [Mixing up 4 stacks](https://youtu.be/z1543RQFdaw?t=23)
*  Less than a second
* 1 call
* 2 items moved

**Review by Jaliborc — APPROVED** [review]
I'll probably improve the distance metric and refactor before release, but thank you for the help!
Great algorithm and detailed explanation!

**Comment by Pierre-Sassoulas** [comment]
Hi @Jaliborc, this suject is fascinating and I'm pretty sure some folks at https://codegolf.stackexchange.com/ would love to think long and hard about it. I had actually started to create a post just a moment before you merged this. I just need to add some example data and instruction so it can be done in multiple languages. Would you mind if I open another merge request later with the best algorithm given in answers ?

**Comment by Jaliborc** [comment]
Sure! But I can't promise how quickly I'll be able to get back to this.

P.S: The repo is currently out of sync with the latest release due to some technical difficulties, I've cleaned your algorithm and improved item distance there.

**Comment by Pierre-Sassoulas** [comment]
Awesome, I'm not super available either and I'm going to wait for the best algorithm anyway. This is such a fancy unconventional problem to solve and it affects a lot of users too, I hope it'll get some traction. Thanks for merging this by the way, I'm proud to be a contributor !

**Pull request merged**
